### PR TITLE
[WIP] Connect to --server (basis for --remote edit and remote TUI)

### PR DESCRIPTION
**Pull request by bfredl**
Both clientserver #8326  and remote TUI #10071 requires connecting to a server address specified on the cmdline. This implements basic client support and a _subset_ of --remote commands ("send", "expr" and nvim specific "lua" subcommand). Both a complete --remote implementation (which will require a bit more arg parsing logic than #8326 has already done) and TUI attachment will be done upon this shared client code.

### CODE REVIEW

File: src/nvim/globals.h [hunk 1]
```diff
@@ -1033,6 +1033,9 @@ EXTERN char_u e_unsupportedoption[] INIT(= N_("E519: Option not supported"));
 EXTERN char_u e_fnametoolong[] INIT(= N_("E856: Filename too long"));
 EXTERN char_u e_float_as_string[] INIT(= N_("E806: using Float as a String"));
 
+EXTERN char_u e_noserver[] INIT(=N_(
+    "E247: couldn't connect to remote server: %s"));
```
Comment (justinmk) [inline src/nvim/globals.h]:
```suggestion
    "E247: failed to connect to remote server: %s"));
```

or `cannot`. Let's avoid contractions in error messages

File: src/nvim/lua/vim.lua [hunk 2]
```diff
@@ -253,6 +253,54 @@ local function fn_index(t, key)
 end
 local fn = setmetatable({}, {__index=fn_index})
 
+local function __rpcrequest(...)
```
Comment (justinmk) [inline src/nvim/lua/vim.lua]: do we need double-underscore? How does it differ from single-underscore?

File: src/nvim/lua/vim.lua [hunk 3]
```diff
@@ -253,6 +253,54 @@ local function fn_index(t, key)
...
+  local f_wait = false
+  local f_tab = false
+  local should_exit = true
+  local command = 'edit '
+
+  local subcmd = string.sub(args[1],10)
+
+  if subcmd == 'send' then
+    __rpcrequest(rcid, 'nvim_input', args[2])
+    -- should we show warning if --server doesn't exist in --send and --expr?
```
Comment (justinmk) [inline src/nvim/lua/vim.lua]: +1

File: src/nvim/lua/vim.lua (hunk 2 above)
Comment (bfredl) [reply src/nvim/lua/vim.lua]: This was just from the  #8326 PR :) Soon it will be `vim.rpcrequest`.

File: src/nvim/main.c [hunk 4]
```diff
@@ -793,6 +893,9 @@ static void command_line_scan(mparm_T *parmp)
           // "--version" give version message
           // "--noplugin[s]" skip plugins
           // "--cmd <cmd>" execute cmd before vimrc
+          // "--remote" open file on remote instance
+          // "--server" name of vim server to send to or name
+          // of server to become
```
Comment (justinmk) [inline src/nvim/main.c]: If we are going to introduce `--server` do we still need `--listen` ?

Comment (bfredl) [reply src/nvim/main.c]: `--server` was just from vim. but as the server names is quite different from vim anyway, I guess we can rename the option as well..

File: test/functional/core/remote_spec.lua [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+local spawn = helpers.spawn
+local set_session = helpers.set_session
+local expect = helpers.expect
+
+describe("remote --server client", function()
+  local server, client, address, pid
+  before_each(function()
+    local nvim_argv = merge_args(helpers.nvim_argv, {'--headless'})
+    server = spawn(nvim_argv)
+    set_session(server)
```
Comment (justinmk) [inline test/functional/core/remote_spec.lua]: I think `merge_args` can be "private", tests don't need to call it directly anymore. `spawn()` and `clear()` merge args implicitly. 

    server = spawn({'--headless'}, true)

Comment (bfredl) [reply test/functional/core/remote_spec.lua]: `clear()` is killing the previous session though. But we could add it an extra flag for that as well.

**Comment by justinmk** [comment]
I wonder if `--remote` is needed at all. 

- Imagine that we claim `-l` to invoke Lua just like `-c` invokes Ex commands. 
- When `--connect` (`--server`) is given, then `-c` and `-l` (and `+`):
  - work on the remote server.
  - return the results of their commands.
- We would need a new `--expr` thing for eval'ing Vimscript expressions.

**Comment by bfredl** [comment]
The idea was that `nvim --remote-xx` without `--server` would use an env var so that it works inside `:term` just like nvr. Tough we could of course add a `--connect [special value]` or `--parent` flag or something for that inside your scheme.

**Comment by justinmk** [comment]
> The idea was that `nvim --remote-xx` without `--server` would use an env var so that it works inside `:term` just like nvr. 

What env var? NVIM_LISTEN_ADRESS  should be cleared in `:term`, it has a conflict of interest: child `nvim` in `:terminal` thinks it is supposed to serve from that (occupied) address.  

For connecting to a parent the plan in https://github.com/neovim/neovim/pull/11009 was to do something like `v:parent` and then use Vimscript/Lua to connect to the parent. nvr emulation with `--remote` seems unimportant.

**Comment by bfredl** [comment]
> What env var? NVIM_LISTEN_ADRESS should be cleared in :term, it has a conflict of interest: child nvim in :terminal thinks it is supposed to serve from that (occupied) address.

Which is exactly why I _didn't_ mention NVIM_LISTEN_ADRESS specifically. The point was that we should use whatever mechanism nvim makes available to child processes, the point was not about precisely what the variable happens to be called.

> For connecting to a parent the plan in #11009 

So we just use `$NVIM` or whatever the bikeshed in that PR turns out to call it.

> For connecting to a parent the plan in #11009 was to do something like v:parent and then use Vimscript/Lua to connect to the parent. nvr emulation with --remote seems unimportant.

So we if have the functionality to connect to a nvim process at the cmdline and do useful stuff with it _and_ the functionally to get the address of the parent nvim, it seem quite natural to use them together. I don't see why we should arbitrarily exclude this functionality, what makes this somehow less "important" than how "important" they are when used separately?

**Comment by justinmk** [comment]
> The idea was that `nvim --remote-xx` without `--server` would use an env var so that it works inside `:term` just like nvr. Tough we could of course add a `--connect [special value]` or `--parent` flag or something for that inside your scheme.

Could `nvim --connect $NVIM` address that use-case? Then we don't need `--remote`. 

Or "worst case", `--remote` is just a boolean arg, which changes the behavior of `-c`, `+`, `-l`.

**Comment by bfredl** [comment]
> Could nvim --connect $NVIM address that use-case? Then we don't need --remote.

What I'm concerned about is the functionality, not its exact name (as I tried to say already in the text you quoted :) `nvim --connect $NVIM` sounds reasonable to have in a shell alias, or something. Do you plan to do `$NVIM` yourself, or should I do it here?

**Comment by justinmk** [comment]
> Do you plan to do `$NVIM` yourself, or should I do it here?

Let me try to wrap up https://github.com/neovim/neovim/pull/11009 next week. Ping me in a couple days if I'm blocking you.

File: src/nvim/lua/vim.lua (hunk 2 above)
Comment (Ethereum7) [reply src/nvim/lua/vim.lua]: 1.

**Comment by Shatur95** [comment]
It would be very nice to have. @justinmk, are you planning to continue work on #11009?

**Comment by bfredl** [comment]
replaced by #17691

**Pull request closed without merging**